File: solutions/problem_1584_min_cost_to_connect_all_points/solution.cpp
/*
 * Problem 1584: Min Cost to Connect All Points
 * ==========================================
 * Difficulty: Medium
 * Tags: Array, Union-Find, Graph Theory, Minimum Spanning Tree
 * Pattern: Prim's MST / Graph
 *
 * Time Complexity:  O(n^2 log n)
 * Space Complexity: O(n^2)
 */

#include <functional>
#include <numeric>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        // Union Find (DSU) - O(n * alpha(n))
        int n = points.size();
        vector<int> parent(n + 1), rnk(n + 1, 0);
        iota(parent.begin(), parent.end(), 0);
        function<int(int)> find = [&](int x) -> int {
            return parent[x] == x ? x : parent[x] = find(parent[x]);
        };
        auto unite = [&](int x, int y) -> bool {
            int px = find(x), py = find(y);
            if (px == py) return false;
            if (rnk[px] < rnk[py]) swap(px, py);
            parent[py] = px;
            if (rnk[px] == rnk[py]) rnk[px]++;
            return true;
        };
        int components = n;
        return components;
    }
};

File: solutions/problem_1584_min_cost_to_connect_all_points/solution.py
"""
Problem 1584: Min Cost to Connect All Points
==========================================
Difficulty: Medium
Tags: Array, Union-Find, Graph Theory, Minimum Spanning Tree
Pattern: Prim's MST / Graph

Time Complexity:  O(n^2 log n)
Space Complexity: O(n^2)
"""

class Solution:
    def minCostConnectPoints(self, points: list[list[int]]) -> int:
        import heapq
        n = len(points)
        visited = set()
        heap = [(0, 0)]
        cost = 0
        while len(visited) < n:
            d, u = heapq.heappop(heap)
            if u in visited:
                continue
            visited.add(u)
            cost += d
            for v in range(n):
                if v not in visited:
                    dist = abs(points[u][0]-points[v][0]) + abs(points[u][1]-points[v][1])
                    heapq.heappush(heap, (dist, v))
        return cost
